terracotta/drivers/relational_base.py

```diff
@@ -0,0 +1,486 @@
+                    Sequence, Tuple, Type, Union)
+
+import numpy as np
+import sqlalchemy as sqla
+import terracotta
+from sqlalchemy.engine.base import Connection
+from terracotta import exceptions
+from terracotta.drivers.base import requires_connection
+from terracotta.drivers.raster_base import RasterDriver
+from terracotta.profile import trace
+
+_ERROR_ON_CONNECT = (
+    'Could not connect to database. Make sure that the given path points '
+    'to a valid Terracotta database, and that you ran driver.create().'
+)
+
+
+def convert_exceptions(error_message: str) -> Callable[..., Any]:
+    def decorator(fun: Callable[..., Any]) -> Callable[..., Any]:
+        @functools.wraps(fun)
+        def inner(self: 'RelationalDriver', *args: Any, **kwargs: Any) -> Any:
+            with convert_exceptions_context(
+                error_message,
+                self.DATABASE_DRIVER_EXCEPTIONS_TO_CONVERT
+            ):
+                return fun(self, *args, **kwargs)
+        return inner
+    return decorator
+
+
+@contextlib.contextmanager
+def convert_exceptions_context(
+    error_message: str,
+    exceptions_to_convert: Union[Type[Exception], Tuple[Type[Exception], ...]]
+) -> Iterator:
+    try:
+        yield
+    except exceptions_to_convert as exception:
+        raise exceptions.InvalidDatabaseError(error_message) from exception
+
+
+class RelationalDriver(RasterDriver, ABC):
+    SQL_DATABASE_SCHEME: str  # The database flavour, eg mysql, sqlite, etc
+    SQL_DRIVER_TYPE: str  # The actual database driver, eg pymysql, sqlite3, etc
+    SQL_KEY_SIZE: int
+    SQL_TIMEOUT_KEY: str
+
+    FILE_BASED_DATABASE: bool = False
+
+    DATABASE_DRIVER_EXCEPTIONS_TO_CONVERT: Tuple[Type[Exception], ...] = (
+        sqla.exc.OperationalError,
+        sqla.exc.InternalError,
+        sqla.exc.ProgrammingError,
+        sqla.exc.InvalidRequestError,
+    )
+
+    SQLA_REAL = functools.partial(sqla.types.Float, precision=8)
+    SQLA_TEXT = sqla.types.Text
+    SQLA_BLOB = sqla.types.LargeBinary
+    _METADATA_COLUMNS: Tuple[Tuple[str, sqla.types.TypeEngine], ...] = (
+        ('bounds_north', SQLA_REAL),
+        ('bounds_east', SQLA_REAL),
+        ('bounds_south', SQLA_REAL),
+        ('bounds_west', SQLA_REAL),
+        ('convex_hull', SQLA_TEXT),
+        ('valid_percentage', SQLA_REAL),
+        ('min', SQLA_REAL),
+        ('max', SQLA_REAL),
+        ('mean', SQLA_REAL),
+        ('stdev', SQLA_REAL),
+        ('percentiles', SQLA_BLOB),
+        ('metadata', SQLA_TEXT)
+    )
+
+    def __init__(self, path: str) -> None:
+        settings = terracotta.get_settings()
+        db_connection_timeout: int = settings.DB_CONNECTION_TIMEOUT
+
+        assert self.SQL_DRIVER_TYPE is not None
+        self._CONNECTION_PARAMETERS = self._parse_connection_string(path)
+        cp = self._CONNECTION_PARAMETERS
+        resolved_path = self._resolve_path(cp.path[1:])  # Remove the leading '/'
```

Comment:
Yes, by its current design, it will always be there. This is part of the reason for the 
```python
        if con_params.scheme == 'file' and cls.FILE_BASED_DATABASE:
            file_connection_string = connection_string[len('file://'):]
            con_params = urlparse.urlparse(f'{cls.SQL_DATABASE_SCHEME}://{file_connection_string}')
```
snippet you stumbled upon just above.

But these quirks and work-arounds are irritating me a bit, and a major part of the reason I want to try to re-implement most of the path/parsing stuff at some point. I will add a more in-depth post regarding my thoughts about that later